core/amber/src/main/python/core/models/tuple.py

```diff
@@ -153,58 +163,76 @@ def as_dict(self) -> Dict[str, AttributeType]:
         return deepcopy(self._field_data)
 
-    def as_key_value_pairs(self) -> List[typing.Tuple[str, AttributeType]]:
+    def as_key_value_pairs(self) -> List[typing.Tuple[str, Field]]:
         return [(k, v) for k, v in self.as_dict().items()]
 
     def get_field_names(self) -> typing.Tuple[str]:
         return tuple(map(str, self._field_data.keys()))
 
-    def get_fields(self, output_field_names=None) -> typing.Tuple[AttributeType]:
+    def get_fields(self, output_field_names=None) -> typing.Tuple[Field]:
         """
         Get values from tuple for selected fields.
         """
         if output_field_names is None:
             output_field_names = self.get_field_names()
         return tuple(self[i] for i in output_field_names)
 
-    def cast_tuple_to_match_schema(self, schema: Schema):
-        # TODO: refactor this function.
+    def finalize(self, schema: Schema) -> None:
+        """
+        Finalizes a Tuple by adding a schema to it. This convert all Fields into the
+        AttributeType defined in the Schema and make the Tuple immutable.
+
+        A Tuple can have no Schema initially. The types of Fields are not restricted.
+        This is to provide the maximum flexibility for users to construct Tuples as
+        they wish. When a Schema is added, the Tuple is finalized to match the Schema.
+
+        :param schema: target Schema to finalize the Tuple.
+        :return:
+        """
+        self.cast_to_schema(schema)
+        self.validate_schema(schema)
+
+    def cast_to_schema(self, schema: Schema) -> None:
+        """
+        Safely cast each field value to match the target schema.
+        If failed, the value will stay not changed.
+
+        This current conducts two kinds of casts:
+            1. cast NaN to None;
+            2. cast any object to bytes (using pickle).
+        :param schema: The target Schema that describes the target AttributeType to
+            cast.
+        :return:
+        """
         for field_name in self.get_field_names():
             try:
+
+                field_value: Field = self[field_name]
+
                 # convert NaN to None to support null value conversion
-                if checknull(self[field_name]):
-                    self[field_name] = None
-                field_value = self[field_name]
-                field = schema.field(field_name)
-                field_type = None if field is None else field.type
-                if field_type == pyarrow.binary():
-                    self[field_name] = b"pickle    " + pickle.dumps(field_value)
-            except Exception:
+                if checknull(field_value):
+                    field_value = None
+
+                if field_value is not None:
+                    field_type = schema.get_attr_type(field_name)
+                    if field_type == AttributeType.BINARY:
+                        field_value = b"pickle    " + pickle.dumps(field_value)
+                self[field_name] = field_value
```

Comment:
Does this mean each field of a tuple will have to go through this assignment? This was not the case before (only `null` and `binary` will). I wonder whether this will have impact on performance.

Reply:
This is a valid point! Such assignments typically will not take much time but it cannot be optimized by Python (see the atteched bytecode).  
```
220     >>   76 LOAD_FAST                3 (field_value)
             78 LOAD_FAST                0 (self)
             80 LOAD_FAST                2 (field_name)
             82 STORE_SUBSCR
             84 POP_BLOCK
             86 JUMP_ABSOLUTE            8
```

I have changed it to do explicit assignments only when needed.